### PR TITLE
issue #38

### PR DESCRIPTION
**Pull request by bazooka70**
https://github.com/dgarage/NBXplorer/issues/38

Added `ExplorerClient.GetDerivations()`
You can query the available derivations via:
HTTP GET v1/cryptos/{cryptoCode}/derivations/getderivations

optional parameters:
```
DerivationStrategyBase strategy = null, 
DerivationFeature? feature = null,
KeyPath keyPath = null, 
Script scriptPubKey = null
```

Examples:

To query all derivations with KeyPaths of "0/0" use:

`var derivations = client.GetDerivations(null, null, new KeyPath("0/0"), null);`

To query for a bitcoin address use for e.g.:

`var derivations = client.GetDerivations(null, null, null, publicAddress.ScriptPubKey);`

Any combination of the optional parameters will use **AND** operator.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
that seems awesome, can you add a small test ?
No need of something fancy, you can hijack one of the existing tests if you want.

**Comment by bazooka70** [comment]
OK, the tests are good. tests passed.

**Comment by NicolasDorier** [comment]
very good! Will test / merge and modify a bit today. :)

**Comment by bazooka70** [comment]
Thanks :) It was good I dived into a fragment of your code. your work is outstanding!

**Comment by NicolasDorier** [comment]
Will do tomorrow actually, there is some small thing I want to modify after merging this.

You added lot's of query capabilities, sadly NBXplorer depends on NoSQL DB which is quite inflexible for queries. The way you are fetching the key is by enumerating all the table. This is very inefficient and would make NBXplorer crash if too many people use it.

The only query you can do is on the scriptPubKey. Because the scriptPubKey is a primary key.

**Comment by bazooka70** [comment]
I know exactly what you mean, but I had no other choices if I wanted to make the query flexible/generic but to load the entire table first. 

Hell, I even wanted to make a `HashSet` for each optional parameter (which is easy to implement), but left the idea to later.

Its a Key-Value db, and problem is that you can't query on the value objects directly.
For each available derivation combination it produces 90 rows. 
So if we have many wallets to track, we load a LOT of rows into memory every time we query even a for a single derivation with a potential to crash. So yes, it's inefficient.

In short, I trust you know better and if you think you can optimize please do so. You can dump the whole idea if you think its not worth the trouble. The last thing I want is a crash because of this feature.

**Comment by NicolasDorier** [comment]
What you wanted to use was `Repository.GetKeyInformations` I don't think you have to modify the Repository at all.

**Comment by bazooka70** [comment]
So you suggest to query only by Scripts[]?

**Comment by bazooka70** [comment]
Something like this?
```csharp
                [HttpGet]
		[Route("cryptos/{cryptoCode}/getkeyinformations")] 
		public IActionResult GetKeyInformations(string cryptoCode,
			[ModelBinder(BinderType = typeof(ScriptsModelBinder))]
			HashSet<Script> scripts
			)
		{
			var network = GetNetwork(cryptoCode);
			var repo = RepositoryProvider.GetRepository(network);
			var keyInformations = repo.GetKeyInformations(scripts.ToArray());
			var result = new List<KeyPathInformation>();
			foreach (var keyInfoByScripts in keyInformations)
			{
				foreach (var keyInfo in keyInfoByScripts.Value)
				{
					result.Add(keyInfo);
				}
			}			
			return Json(result);
		}
```

**Comment by NicolasDorier** [comment]
Fix:

```
[Route("cryptos/{cryptoCode}/scripts/{script}")] 
```

The paramater should be one `Script script`. The ModelBinder should just read it as hex and deserialize it into a Script.

**Comment by bazooka70** [comment]
Can the `repo.GetKeyInformations()` return more than a single element for one `Script`? 
I read this comment:

> // Because xpub are mutable (several xpub map to same script)
> // an attacker could generate lot's of xpub mapping to the same script
> // and this would blow up here. This we take only 5 results max.

And I'm not sure I follow how is that possible...

**Comment by NicolasDorier** [comment]
Yes one can generate infinite `xpub` which create all same script. The reason is that not all info inside one xpub it relevant to derivation.

IMHO you should just returns either the array or the first result after having stripped the ExtPubKey from the KeyInformation so consumer's can't use it and can't shoot themselves in the foot by relying on it.

**Comment by bazooka70** [comment]
I still don't get it. how can you have several KeyInformation for a single Script which was derived from a Keypath and suppose to be unique for the cryptocoin/network?
Can you give me a real example?

> after having stripped the ExtPubKey from the KeyInformation...

But this is exactly what I wanted to return in the first place.

Or you mean to say that there can be a situation where you have derivation of:
`tpubD6NzVbkrYhZ4XtRwac...-[legacy]`
`tpubD6NzVbkrYhZ4XtRwac...-[p2sh]`
etc..
which map to the same Script?

**Comment by bazooka70** [comment]
This is what I came up with:

```csharp
                [HttpGet]
		[Route("cryptos/{cryptoCode}/scripts/{script}")] 
		public IActionResult GetKeyInformations(string cryptoCode,
			[ModelBinder(BinderType = typeof(ScriptModelBinder))] Script script)
		{
			var network = GetNetwork(cryptoCode);
			var repo = RepositoryProvider.GetRepository(network);
			var scripts = new Script[] { script };
			var keyInformations = repo.GetKeyInformations(scripts);			
			var result = new List<KeyPathInformation>();			
			foreach (var keyInfoByScripts in keyInformations)
			{
				foreach (var keyInfo in keyInfoByScripts.Value)
				{
					result.Add(keyInfo);
				}
			}
			if (result.Count == 0)							
				return NotFound();			
			return Json(result.ToArray());
		}
```

ScriptModelBinder:

```csharp
namespace NBXplorer.ModelBinders
{
	public class ScriptModelBinder : IModelBinder
	{
		#region IModelBinder Members

		public Task BindModelAsync(ModelBindingContext bindingContext)
		{
			if (!typeof(Script).GetTypeInfo().IsAssignableFrom(bindingContext.ModelType))
			{
				return Task.CompletedTask;
			}

			ValueProviderResult val = bindingContext.ValueProvider.GetValue(
				bindingContext.ModelName);
			if (val == null)
			{
				return Task.CompletedTask;
			}

			string key = val.FirstValue as string;
			if (key == null)
			{
				bindingContext.Model = null;
				return Task.CompletedTask;
			}

			var value = new Script(Encoders.Hex.DecodeData(key));

			bindingContext.Result = ModelBindingResult.Success(value);
			return Task.CompletedTask;
		}

		#endregion
	}	
}
```

**Comment by NicolasDorier** [comment]
Seems good!

```
xpub661MyMwAqRbcFrVYKi4Gwhhx3Qhf8xMMjTN6RzdBFC7NsGto96G3dPy2fssQoaytHbj7qR7WBs5GFTDvPHuSLKgER36RrSdnpBX2KCsS9Qg
```

and 

```
xpub69mdgvyDG2wbyPkUPgyoZyEddrXYYQaFsFcuBiNWb4qBj6nRDfGjCm8EB2R6HTnpAzbR7Z2vhwVPcQCY81S2T9t6aLkHF7ECTeBvRECTyk4
```

Map to same derivations. Because the depth value is not implicated in the derivation but influence the base58 format.

**Comment by bazooka70** [comment]
Ahhh? I need to check the above this is new to me...
In any case I think I will return an array so everyone but me can shoot themself in the leg :-p
I cracked the code and it works fine.
Do I open a new PR or delete the shit I did or what?

**Comment by NicolasDorier** [comment]
as you want, you can make new PR or update this one by pushing your work to your repo/branch.
